cartography/intel/aws/permission_relationships.py

```diff
@@ -0,0 +1,200 @@
+import logging
+import os
+import re
+
+import yaml
+
+from cartography.graph.statement import GraphStatement
+
+logger = logging.getLogger(__name__)
+
+
+def evaluate_clause(clause, match):
+    # AWS [not]actions and [not]resources can use
+    # variable length wildcards (*)
+    # fixed length wildcards (?)
+
+    result = compile_regex(clause).fullmatch(match)
+    return result is not None
+
+
+def evaluate_statment_clause_for_permission(statement, clause_name, match, missing_clause_return=False):
+    if clause_name not in statement:
+        return missing_clause_return
+    for clause in statement[clause_name]:
+        if evaluate_clause(clause, match):
+            return True
+    return False
+
+
+def evaluate_statements_for_permission(statements, permission, resource_arn):
+    allowed = False
+    for statement in statements:
+        if not evaluate_statment_clause_for_permission(statement, "notaction", permission):
+            if evaluate_statment_clause_for_permission(statement, "action", permission, missing_clause_return=True):
+                if evaluate_statment_clause_for_permission(statement, "resource", resource_arn):
+                    if not evaluate_statment_clause_for_permission(statement, "notresource", resource_arn):
+                        return True
+
+    return allowed
+
+
+def evaluate_policy_for_permission(statements, permissions, resource_arn):
+    # AWS Policy evaluation reference
+    # https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html
+    # returns (allowed_by_policy, explicitly_denied_by_policy)
+    # return cases
+    # (True, False) - The policy allows the action
+    # (False, False) - The poliy does not allow the action, but also doesn't explicitly deny it
+    # (False, True) - The policy specifically denies the action. There is no need to evaluate other policies
+
+    allow_statements = [s for s in statements if s["effect"] == "Allow"]
+    deny_statements = [s for s in statements if s["effect"] == "Deny"]
+    for permission in permissions:
+        if evaluate_statements_for_permission(deny_statements, permission, resource_arn):
+            return False, True
```

Comment:
From reading this function by itself, It isn't clear why returning `False, True` here represents "The policy does not allow the action, but also doesn't explicitly deny it" as explained above.

We should reconsider this set of function interfaces to make them clearer. The code should be self-documenting in that it should be clear why `False, True` is being returned from here as opposed to any other truth pair.